### ISSUE TITLE
[Q-Learning订单执行优化] | Q-Learning Order Execution

### ISSUE THREAD

**Issue by Kaoruha** [Kaoruha/GinkGO#4511]
**标签**: quantitative-strategy
### 摘要
使用Q-Learning强化学习算法优化大额订单执行策略，在限价单与市价单之间动态决策，最小化市场冲击成本。

### 算法描述

**核心原理**

订单执行问题的本质是在**执行速度**与**市场冲击**之间寻找最优平衡。过于激进的执行策略会造成较大的市场冲击，过于保守则会暴露价格风险。

**状态空间（State）**
$$s_t = (P_t, \sigma_t, V_t, t, T, \text{remaining\_shares})$$

其中：
- $P_t$：当前价格
- $\sigma_t$：价格波动率
- $V_t$：交易量
- $t$：当前时间
- $T$：总执行期限
- $\text{remaining\_shares}$：剩余股份数

**动作空间（Action）**
$$a_t \in \{0, 1, 2, ..., N\}$$

表示第t时刻提交的订单量占剩余股份的比例：
$$a_t = k \cdot \text{remaining\_shares}, \quad k \in [0, 1]$$

**奖励函数（Reward）**
$$r_t = -\text{market_impact}(V_t) - \text{opportunity_cost}(\text{unexecuted})$$

**Q-Learning更新公式**
$$Q(s_t, a_t) \leftarrow Q(s_t, a_t) + \alpha [r_t + \gamma \max_{a'} Q(s_{t+1}, a') - Q(s_t, a_t)]$$

**市场冲击模型**

线性冲击模型：
$$\text{impact} = \lambda \cdot \frac{v}{V_{daily}} \cdot \sigma$$

其中$\lambda$为冲击系数，$v$为订单量，$V_{daily}$为日均交易量。

### Python代码示例

```python
"""
Q-Learning订单执行优化策略
使用强化学习算法优化大额订单执行，最小化市场冲击成本
"""
import numpy as np
import pandas as pd
from dataclasses import dataclass
from typing import Dict, List, Tuple, Optional
from collections import defaultdict
import random


@dataclass
class MarketState:
    """市场状态"""
    price: float
    volatility: float
    volume: float
    bid_ask_spread: float
    remaining_shares: int
    time_step: int
    total_steps: int


@dataclass
class OrderExecutionResult:
    """订单执行结果"""
    total_cost: float
    market_impact: float
    implementation_shortfall: float
    fill_rate: float
    avg_price: float


class QLearningExecutor:
    """Q-Learning订单执行器"""
    
    def __init__(
        self,
        state_dim: int = 6,
        action_dim: int = 11,
        learning_rate: float = 0.1,
        discount_factor: float = 0.95,
        epsilon: float = 0.1
    ):
        self.state_dim = state_dim
        self.action_dim = action_dim  # 0-10, 代表执行比例 0%-100%
        self.lr = learning_rate
        self.gamma = discount_factor
        self.epsilon = epsilon
        
        # Q表
        self.Q = defaultdict(lambda: np.zeros(action_dim))
        
        # 超参数
        self.impact_coef = 0.1  # 冲击系数
        self.opportunity_cost_rate = 0.001  # 机会成本率
    
    def discretize_state(self, state: MarketState) -> Tuple:
        """状态离散化"""
        # 价格变化离散化
        price_bin = int(np.clip(state.price / 100, 0, 19))  # 假设价格范围
        
        # 波动率离散化
        vol_bin = int(np.clip(state.volatility * 100, 0, 9))
        
        # 交易量离散化
        vol_level_bin = int(np.clip(np.log10(state.volume + 1), 0, 9))
        
        # 剩余比例离散化
        remaining_ratio = state.remaining_shares / (state.total_steps * 1000)  # 假设每步最大1000股
        remaining_bin = int(np.clip(remaining_ratio * 10, 0, 9))
        
        # 时间步离散化
        time_bin = int(state.time_step / (state.total_steps / 10))
        
        return (price_bin, vol_bin, vol_level_bin, remaining_bin, time_bin)
    
    def choose_action(self, state: MarketState, training: bool = True) -> int:
        """选择动作（epsilon-greedy）"""
        state_key = self.discretize_state(state)
        
        if training and random.random() < self.epsilon:
            return random.randint(0, self.action_dim - 1)
        
        return int(np.argmax(self.Q[state_key]))
    
    def execute_shares(self, state: MarketState, action: int) -> Tuple[float, MarketState]:
        """执行交易并计算成本"""
        # 动作比例
        execute_ratio = action / (self.action_dim - 1)
        execute_shares = int(state.remaining_shares * execute_ratio)
        
        if execute_shares <= 0:
            return 0.0, state
        
        # 市场冲击成本
        daily_volume = state.volume * state.total_steps  # 估算日成交量
        impact = self.impact_coef * (execute_shares / daily_volume) * state.volatility * state.price
        
        # 更新状态
        new_remaining = max(0, state.remaining_shares - execute_shares)
        new_state = MarketState(
            price=state.price * (1 + np.random.normal(0, state.volatility / np.sqrt(252))),
            volatility=state.volatility,
            volume=state.volume,
            bid_ask_spread=state.bid_ask_spread,
            remaining_shares=new_remaining,
            time_step=state.time_step + 1,
            total_steps=state.total_steps
        )
        
        total_cost = impact + self.opportunity_cost_rate * execute_shares * state.price
        return total_cost, new_state
    
    def update_q(
        self,
        state: MarketState,
        action: int,
        reward: float,
        next_state: MarketState
    ):
        """更新Q值"""
        state_key = self.discretize_state(state)
        next_state_key = self.discretize_state(next_state)
        
        current_q = self.Q[state_key][action]
        next_max_q = np.max(self.Q[next_state_key])
        
        # Q-Learning更新
        new_q = current_q + self.lr * (reward + self.gamma * next_max_q - current_q)
        self.Q[state_key][action] = new_q
    
    def train(
        self,
        episodes: int = 1000,
        initial_shares: int = 10000,
        total_steps: int = 20
    ):
        """训练Q-Learning代理"""
        training_history = []
        
        for episode in range(episodes):
            # 初始化状态
            state = MarketState(
                price=100.0,
                volatility=0.02,
                volume=1000000,
                bid_ask_spread=0.01,
                remaining_shares=initial_shares,
                time_step=0,
                total_steps=total_steps
            )
            
            total_cost = 0.0
            is_terminal = False
            
            while not is_terminal and state.remaining_shares > 0:
                # 选择动作
                action = self.choose_action(state, training=True)
                
                # 执行
                cost, next_state = self.execute_shares(state, action)
                
                # 计算奖励（负成本）
                reward = -cost
                
                # 检查终止条件
                if next_state.time_step >= next_state.total_steps or next_state.remaining_shares <= 0:
                    is_terminal = True
                    # 未完成惩罚
                    opportunity_penalty = -self.opportunity_cost_rate * next_state.remaining_shares * next_state.price
                    reward += opportunity_penalty
                
                # 更新Q表
                self.update_q(state, action, reward, next_state)
                
                total_cost += cost
                state = next_state
            
            training_history.append(total_cost)
            
            if episode % 100 == 0:
                avg_cost = np.mean(training_history[-100:])
                print(f"Episode {episode}: Avg Cost = {avg_cost:.2f}")
    
    def execute_order(self, initial_shares: int, total_steps: int) -> OrderExecutionResult:
        """使用训练好的策略执行订单"""
        state = MarketState(
            price=100.0,
            volatility=0.02,
            volume=1000000,
            bid_ask_spread=0.01,
            remaining_shares=initial_shares,
            time_step=0,
            total_steps=total_steps
        )
        
        total_cost = 0.0
        total_executed = 0
        executed_prices = []
        
        while state.remaining_shares > 0 and state.time_step < state.total_steps:
            action = self.choose_action(state, training=False)
            
            execute_ratio = action / (self.action_dim - 1)
            execute_shares = int(state.remaining_shares * min(execute_ratio * 2, 1.0))  # 加速执行
            
            if execute_shares > 0:
                daily_volume = state.volume * state.total_steps
                impact = self.impact_coef * (execute_shares / daily_volume) * state.volatility * state.price
                
                executed_prices.append(state.price)
                total_cost += impact
                total_executed += execute_shares
                
                state = MarketState(
                    price=state.price * (1 + np.random.normal(0, state.volatility / np.sqrt(252))),
                    volatility=state.volatility,
                    volume=state.volume,
                    bid_ask_spread=state.bid_ask_spread,
                    remaining_shares=max(0, state.remaining_shares - execute_shares),
                    time_step=state.time_step + 1,
                    total_steps=state.total_steps
                )
            else:
                break
        
        avg_price = np.mean(executed_prices) if executed_prices else state.price
        remaining_cost = self.opportunity_cost_rate * state.remaining_shares * state.price
        
        return OrderExecutionResult(
            total_cost=total_cost + remaining_cost,
            market_impact=total_cost,
            implementation_shortfall=total_cost / (initial_shares * state.price),
            fill_rate=total_executed / initial_shares if initial_shares > 0 else 0,
            avg_price=avg_price
        )


class MarketSimulator:
    """市场模拟器"""
    
    def __init__(
        self,
        initial_price: float = 100.0,
        volatility: float = 0.02,
        daily_volume: float = 1000000
    ):
        self.initial_price = initial_price
        self.volatility = volatility
        self.daily_volume = daily_volume
        self.price = initial_price
    
    def reset(self):
        """重置市场"""
        self.price = self.initial_price
    
    def step(self, executed_shares: int = 0) -> Dict:
        """模拟一步市场"""
        # 价格随机游走
        price_change = np.random.normal(0, self.volatility / np.sqrt(252))
        self.price *= (1 + price_change)
        
        # 如果有执行，更新价格（冲击）
        if executed_shares > 0:
            impact = 0.0001 * executed_shares / (self.daily_volume / 10000)
            self.price *= (1 + impact)
        
        return {
            'price': self.price,
            'volume': self.daily_volume * np.random.uniform(0.8, 1.2)
        }


class OptimalExecutionStrategy:
    """最优执行策略（对比基准）"""
    
    def __init__(self):
        self.impact_coef = 0.1
    
    defVWAP_strategy(
        self,
        shares: int,
        steps: int,
        market: MarketSimulator
    ) -> List[Dict]:
        """VWAP策略"""
        market.reset()
        executions = []
        shares_per_step = shares // steps
        
        for step in range(steps):
            market_info = market.step(executed_shares=shares_per_step)
            
            executions.append({
                'step': step,
                'price': market_info['price'],
                'shares': shares_per_step,
                'remaining': shares - (step + 1) * shares_per_step
            })
        
        return executions
    
    def POV_strategy(
        self,
        shares: int,
        steps: int,
        market: MarketSimulator,
        participation_rate: float = 0.1
    ) -> List[Dict]:
        """POV（Percentage of Volume）策略"""
        market.reset()
        executions = []
        remaining = shares
        
        for step in range(steps):
            market_info = market.step()
            target_shares = int(market_info['volume'] * participation_rate)
            execute_shares = min(target_shares, remaining)
            
            market.step(executed_shares=execute_shares)
            
            executions.append({
                'step': step,
                'price': market_info['price'],
                'shares': execute_shares,
                'remaining': remaining - execute_shares
            })
            
            remaining -= execute_shares
            if remaining <= 0:
                break
        
        return executions


def evaluate_strategy(
    strategy_name: str,
    executor: QLearningExecutor,
    market: MarketSimulator,
    shares: int = 10000,
    steps: int = 20
) -> Dict:
    """评估策略性能"""
    market.reset()
    
    if strategy_name == "Q-Learning":
        result = executor.execute_order(shares, steps)
        return {
            'strategy': strategy_name,
            'total_cost': result.total_cost,
            'market_impact': result.market_impact,
            'impl_shortfall': result.implementation_shortfall,
            'fill_rate': result.fill_rate
        }
    else:
        # VWAP基准
        opt_exec = OptimalExecutionStrategy()
        executions = opt_exec.VWAP_strategy(shares, steps, market)
        
        total_cost = 0
        for ex in executions:
            daily_vol = market.daily_volume
            impact = 0.0001 * ex['shares'] / (daily_vol / 10000) * market.volatility * ex['price']
            total_cost += impact
        
        return {
            'strategy': strategy_name,
            'total_cost': total_cost,
            'fill_rate': 1.0
        }


if __name__ == '__main__':
    print("=== Q-Learning订单执行优化训练 ===")
    
    executor = QLearningExecutor(
        learning_rate=0.1,
        discount_factor=0.95,
        epsilon=0.2
    )
    
    # 训练
    executor.train(episodes=500, initial_shares=10000, total_steps=20)
    
    # 评估
    print("\n=== 策略评估 ===")
    market = MarketSimulator(initial_price=100, volatility=0.02, daily_volume=1000000)
    
    for strategy in ["Q-Learning", "VWAP"]:
        metrics = evaluate_strategy(strategy, executor, market, shares=10000, steps=20)
        print(f"\n{metrics['strategy']}:")
        print(f"  总成本: {metrics['total_cost']:.2f}")
        print(f"  市场冲击: {metrics.get('market_impact', 'N/A')}")
        print(f"  实现缺口: {metrics.get('impl_shortfall', 'N/A')}")
        print(f"  成交率: {metrics['fill_rate']:.2%}")
```

### 数据来源
- Reddit: r/quant, r/algorithmictrading
- 小红书: 量化投资
- 知乎: 量化交易
